[Console] Added suggestions for missing packages

przemyslaw-bogusz opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT

Currently, when someone runs one of the most common commands, e.g. `server:run`, but does not have a required package installed, they will get a general **'There are no commands defined...'** message.

This commit adds a more useful message, informing the user about a package that might be missing and suggesting a command that should be run in order to install it, e.g. `composer require symfony/web-server-bundle --dev`.

ro0NL commented:
i'd prefer something like we did for twig: https://github.com/symfony/symfony/blob/master/src/Symfony/Bridge/Twig/UndefinedCallableHandler.php

i think its more straightforward to handle it in e.g. a catch-all event listener for unknown commands, opposed to tracking a list on high level (console application).

przemyslaw-bogusz commented:
I had some doubts adding it to the main application, since it is widely used as a standalone component, but I wanted the suggestions to be available for some, who started his project with a symfony/skeleton (so without the FrameworkBundle) and tries to use features, that are not installed.

By the way, can you tell me, why my CI checks failed? Did I forget about something in my PR?

przemyslaw-bogusz commented:
My mistake. I confused FrameworkBundle with SensioFrameworkExtraBundle. I will rework the PR.

But my other question about the CI checks is still valid :)

przemyslaw-bogusz committed:
[Console] Added suggestions for missing packages

przemyslaw-bogusz commented:
I went with the suggestion from @ro0NL and created an event subscriber. I can always switch to a try/catch in [FrameworkBundle Console Application](https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/Console/Application.php).

mbabker commented on src/Symfony/Bundle/FrameworkBundle/EventListener/SuggestMissingPackageSubscriber.php:
```diff
@@ -0,0 +1,67 @@
+ */
+final class SuggestMissingPackageSubscriber implements EventSubscriberInterface
+{
+    const SUGGESTED_PACKAGES = [
```
Mark private?  I don't think there's a reason this needs to be publicly accessible.

przemyslaw-bogusz replied:
Changed into private static property, as it seemed to be in line with other Symfony's classes.

mbabker commented on src/Symfony/Bundle/FrameworkBundle/EventListener/SuggestMissingPackageSubscriber.php:
```diff
@@ -0,0 +1,67 @@
+                if (isset(self::SUGGESTED_PACKAGES[$namespace])) {
+                    $suggestion = self::SUGGESTED_PACKAGES[$namespace];
+                    $message = sprintf("It seems that you are trying to run a command from '%s' bundle/package, but you do not have it installed. You may try installing the missing bundle/package by running:\n\n    composer require %s", $suggestion[0], $suggestion[1]);
+                    throw new NamespaceNotFoundException($message);
```
`$event->setError()` instead of throwing a new Exception, this way other error event listeners can still be actioned.

przemyslaw-bogusz replied:
Done

chalasr commented on src/Symfony/Bundle/FrameworkBundle/EventListener/SuggestMissingPackageSubscriber.php:
```diff
@@ -0,0 +1,72 @@
+
+                if (isset(self::$suggestedPackages[$namespace])) {
+                    $suggestion = self::$suggestedPackages[$namespace];
+                    $message = sprintf("It seems that you are trying to run a command from '%s' bundle/package, but you do not have it installed. You may try installing the missing bundle/package by running:\n\n    composer require %s", $suggestion[0], $suggestion[1]);
```
Exceptions message should never be formatted, `Application::renderException()` already takes care of the output. 
Here is a proposal that is consistent with existing DX error messages:
```php
$message = sprintf('%s You may be looking for a command provided by the %s which is not installed currently. Try running "composer require %s".', $e->getMessage(), $suggestion[0], $suggestion[1]);
```

Which gives:
> There are no commands defined in the "make" namespace. You may be looking for a command provided by the MakerBundle which is not installed currently, try running "composer require symfony/maker-bundle --dev".

przemyslaw-bogusz replied:
By formatted you mean adding new lines?

chalasr replied:
Nevermind

chalasr commented on src/Symfony/Bundle/FrameworkBundle/EventListener/SuggestMissingPackageSubscriber.php:
```diff
@@ -0,0 +1,72 @@
+            'symfony/maker-bundle --dev',
+        ],
+        'server' => [
+            'Symfony Web Server',
```
should be `WebServerBundle`

przemyslaw-bogusz replied:
Done

chalasr commented on src/Symfony/Bundle/FrameworkBundle/EventListener/SuggestMissingPackageSubscriber.php:
```diff
@@ -0,0 +1,72 @@
+
+    public function onConsoleError(ConsoleErrorEvent $event)
+    {
+        if ($event->getError() instanceof NamespaceNotFoundException) {
```
Please use one guard clause instead:

```php
if (!$event->getError() instanceof NamespaceNotFoundException || !$namespace = $event->getInput()->getFirstArgument() || !isset(self::SUGGESTED_PACKAGES[$namespace])) {
    return;
}

przemyslaw-bogusz replied:
I switched to guard clause but it's a little different since I modified the whole logic

chalasr reviewed (changes requested):
We will need some tests

przemyslaw-bogusz commented on src/Symfony/Bundle/FrameworkBundle/EventListener/SuggestMissingPackageSubscriber.php:
```diff
@@ -0,0 +1,67 @@
+            'MakerBundle',
+            'symfony/maker-bundle --dev',
+        ],
+        'server' => [
```
Done. Analogically for `doctrine:`.

przemyslaw-bogusz commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/console.xml:
```diff
@@ -13,6 +13,11 @@
         </service>
 
+        <service id="console.suggest_missing_package_subscriber"
+                 class="Symfony\Bundle\FrameworkBundle\EventListener\SuggestMissingPackageSubscriber">
```
Done

przemyslaw-bogusz commented on src/Symfony/Bundle/FrameworkBundle/EventListener/SuggestMissingPackageSubscriber.php:
```diff
@@ -0,0 +1,67 @@
+        ],
+    ];
+
+    public function onConsoleError(ConsoleErrorEvent $event)
```
**Coding Standards** indicate

>  use `return;` when the function returns `void` values

Should I also add it?

chalasr replied:
Looks like we need to upgrade our coding standards :)

przemyslaw-bogusz commented on src/Symfony/Bundle/FrameworkBundle/EventListener/SuggestMissingPackageSubscriber.php:
```diff
@@ -0,0 +1,67 @@
+        }
+    }
+
+    public static function getSubscribedEvents()
```
Done

przemyslaw-bogusz commented on src/Symfony/Bundle/FrameworkBundle/EventListener/SuggestMissingPackageSubscriber.php:
```diff
@@ -0,0 +1,72 @@
+     *
+     * @var array
+     */
+    private static $suggestedPackages = [
```
Fixed along with other things to make **fabbot** happy

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/EventListener/SuggestMissingPackageSubscriber.php:
```diff
@@ -0,0 +1,83 @@
+ *
+ * @author Przemysław Bogusz <przemyslaw.bogusz@tubotax.pl>
+ */
+final class SuggestMissingPackageSubscriber implements EventSubscriberInterface
```
since this provides a behavior and we don't want to provide more (ie no design blocks), I'd suggest making this class `@internal`

przemyslaw-bogusz replied:
Done

chalasr commented:
@przemyslaw-bogusz Could you please take @nicolas-grekas's comment into account? This looks good to me.

przemyslaw-bogusz commented:
@chalasr @nicolas-grekas All done. I've also updated changelog

fabpot commented:
@przemyslaw-bogusz Can you rebase this pull request? I see that you merged instead of rebase and that prevents us from merging. Thank you.

przemyslaw-bogusz commented:
@fabpot Done.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/EventListener/SuggestMissingPackageSubscriber.php:
```diff
@@ -0,0 +1,85 @@
+            return;
+        }
+
+        $message = sprintf("%s\n\nYou may be looking for a command provided by the \"%s\" which is currently not installed. Try running \"composer require %s\".", $error->getMessage(), $suggestion[0], $suggestion[1]);
```
I'd suggest removing the "the" eg "the Doctrine ORM" doesn't sound that nice.

przemyslaw-bogusz replied:
Done

przemyslaw-bogusz replied:
It seems it's "the Doctrine ORM" after all
https://symfony.com/doc/current/doctrine.html

przemyslaw-bogusz replied:
`the` is back again

nicolas-grekas reviewed (approved):
(with minor comment)

fabpot commented on src/Symfony/Bundle/FrameworkBundle/EventListener/SuggestMissingPackageSubscriber.php:
```diff
@@ -0,0 +1,85 @@
+
+    public function onConsoleError(ConsoleErrorEvent $event): void
+    {
+        if (!$event->getError() instanceof CommandNotFoundException || empty($input = $event->getInput()->getFirstArgument())) {
```
Can you remove the usage of `empty`. We tend to avoid using it as many "valid" strings are "empty" like `0` for instance.

przemyslaw-bogusz replied:
Done

fabpot commented on src/Symfony/Bundle/FrameworkBundle/EventListener/SuggestMissingPackageSubscriber.php:
```diff
@@ -0,0 +1,85 @@
+
+        $error = $event->getError();
+
+        if (!empty($error->getAlternatives()) && !$exact) {
```
Same here, you can just remove `empty`

przemyslaw-bogusz replied:
This one is for an array. I can use `count`, if it's preferable. If I don't check that the array has any elements, the method will not return a proper message.

For example, on running `php bin/console doctrine` it will just state `Command "doctrine" is not defined.` instead of `Command "doctrine" is not defined. You may be looking for a command provided by "Doctrine ORM" which is currently not installed. Try running "composer require symfony/orm-pack". `

nicolas-grekas replied:
` if ($error->getAlternatives() && !$exact) {` would just work, no need for anything around.

przemyslaw-bogusz replied:
Done

chalasr commented on src/Symfony/Bundle/FrameworkBundle/EventListener/SuggestMissingPackageSubscriber.php:
```diff
@@ -0,0 +1,85 @@
+            '_default' => ['MakerBundle', 'symfony/maker-bundle --dev'],
+        ],
+        'server' => [
+            'dump' => ['VarDumper Component', 'symfony/var-dumper --dev'],
```
I would keep `the` for this one

> You may be looking for a command provided by the VarDumper Component

przemyslaw-bogusz replied:
@nicolas-grekas @chalasr  I checked in the docs. These have `the`:
the Doctrine ORM
https://symfony.com/doc/current/doctrine.html
the MakerBundle
https://symfony.com/doc/current/bundles/SymfonyMakerBundle/index.html
the VarDumper
https://symfony.com/doc/current/components/var_dumper.html
Installing the Web Server Bundle
https://symfony.com/doc/current/setup/built_in_web_server.html

So maybe it's a bad idea to keep `the` for all, after all?

chalasr replied:
Readding `the` for all suggestions looks fine to me since it is consistent with our docs

przemyslaw-bogusz replied:
done

chalasr commented on src/Symfony/Bundle/FrameworkBundle/Tests/Console/ApplicationTest.php:
```diff
@@ -226,6 +229,34 @@ public function testRunOnlyWarnsOnUnregistrableCommandAtTheEnd()
+    public function testSuggestingPackagesWithPartialMatchAndAlternatives()
+    {
+        $result = $this->createEventForSuggestingPackages('server', ['server:run']);
+        $this->assertNotRegExp('/You may be looking for a command provided by the/', $result);
```
`the` should be removed I guess

przemyslaw-bogusz replied:
Done. I was fixing it when you commented

ro0NL commented on src/Symfony/Bundle/FrameworkBundle/EventListener/SuggestMissingPackageSubscriber.php:
```diff
@@ -0,0 +1,85 @@
+            return;
+        }
+
+        $input = explode(':', $input);
```
`[$namespace, $command] = explode(':', $input) + [1 => ''];` saves a few bits.

przemyslaw-bogusz replied:
done

ro0NL commented on src/Symfony/Bundle/FrameworkBundle/EventListener/SuggestMissingPackageSubscriber.php:
```diff
@@ -0,0 +1,85 @@
+            return;
+        }
+
+        if (isset(self::PACKAGES[$namespace][$command]) && '' !== $command) {
```
no real need to check for `'' !== $command` here.

przemyslaw-bogusz replied:
done

ro0NL commented on src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md:
```diff
@@ -16,6 +16,7 @@ CHANGELOG
  * Added php ini session options `sid_length` and `sid_bits_per_character` 
    to the `session` section of the configuration
  * Added support for Translator paths, Twig paths in translation commands.
+ * Added `SuggestMissingPackageSubscriber` for suggesting missing packages when using console commands
```
not sure we need  to advertise this internal class =/

przemyslaw-bogusz replied:
done

przemyslaw-bogusz commented:
Too many reviews in a short period of time and things got a little messy for a moment. I've implemented all suggestions the code, fixed the tests. The one Travis CI test that fails, failed before and I guess it has nothing to do with my code directly.

The only thing left to decide is whether to put `the` back in the exception message by default, or add it to selected messages. I'm in favor of `the` by default.

fabpot commented:
Thank you @przemyslaw-bogusz.
